path/dynamic/dstarlite.go

```diff
@@ -0,0 +1,462 @@
+// Copyright ©2014 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dynamic
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/concrete"
+	"github.com/gonum/graph/path"
+)
+
+// DStarLite implements the D* Lite dynamic re-planning path search algorithm.
+//
+//  doi:10.1109/tro.2004.838026 and ISBN:0-262-51129-0 pp476-483
+//
+type DStarLite struct {
+	s, t *dStarLiteNode
+	last *dStarLiteNode
+
+	world       MutableDirected
+	queue       dStarLiteQueue
+	keyModifier float64
+
+	weight    graph.WeightFunc
+	heuristic path.Heuristic
+}
+
+// MutableDirected is a mutable directed graph that can return nodes identified
+// by id number.
+type MutableDirected interface {
+	graph.MutableDirected
+	Node(id int) graph.Node
+}
+
+// NewDStarLite returns a new DStarLite planner for the path from s to t in g using the
+// heuristic h. The mutable directed graph world is used to store shortest path information
+// during path planning.
+//
+// If h is nil, the DStarLite will use the g.HeuristicCost method if g implements
+// path.HeuristicCoster, falling back to path.NullHeuristic otherwise. If the graph does not
+// implement graph.Weighter, graph.UniformCost is used. NewDStarLite will panic if g has
+// a negative edge weight.
+func NewDStarLite(s, t graph.Node, g graph.Graph, h path.Heuristic, world MutableDirected) *DStarLite {
+	/*
+	   procedure Initialize()
+	   {02”} U = ∅;
+	   {03”} k_m = 0;
+	   {04”} for all s ∈ S rhs(s) = g(s) = ∞;
+	   {05”} rhs(s_goal) = 0;
+	   {06”} U.Insert(s_goal, [h(s_start, s_goal); 0]);
+	*/
+
+	d := &DStarLite{
+		s: &dStarLiteNode{Node: s, rhs: math.Inf(1), g: math.Inf(1)},
+		t: &dStarLiteNode{Node: t, g: math.Inf(1)},
+
+		world: world,
+
+		heuristic: h,
+	}
+
+	/*
+		procedure Main()
+		{29”} s_last = s_start;
+		{30”} Initialize();
+	*/
+	d.last = d.s
+
+	if g, ok := g.(graph.Weighter); ok {
+		d.weight = g.Weight
+	} else {
+		d.weight = graph.UniformCost
+	}
+	if d.heuristic == nil {
+		if g, ok := g.(path.HeuristicCoster); ok {
+			d.heuristic = g.HeuristicCost
+		} else {
+			d.heuristic = path.NullHeuristic
+		}
+	}
+
+	d.queue.indexOf = make(map[int]int)
+	d.queue.insert(d.t, key{h(s, t), 0})
+
+	for _, n := range g.Nodes() {
+		switch n.ID() {
+		case d.s.ID():
+			world.AddNode(d.s)
+		case d.t.ID():
+			world.AddNode(d.t)
+		default:
+			world.AddNode(&dStarLiteNode{Node: n, rhs: math.Inf(1), g: math.Inf(1)})
+		}
+	}
+	for _, u := range world.Nodes() {
+		for _, v := range g.From(u) {
+			w := d.weight(g.Edge(u, v))
```

Comment:
Yes, that's fine; edge lookup is based on node ID, so so long as the IDs agree there is no problem and that is the case here.